File: src/protocol/engine.ts
import type { FileCapability, FileKind } from "./file-system"

/** 引擎只描述展示需求；具体 React 组件由 UI 层按 engineId 另行注册。 */
export type EngineLayout = "padded" | "fill"

export type EngineAccess = "read-only" | "read-write"

export type EngineMatchPropertyValue = string | number | boolean | null

export type EngineMatcher = Readonly<{
  /** 不指定即接受 file 与 directory。 */
  kinds?: readonly FileKind[]
  /** MIME 模式；支持精确值、类型通配、全通配以及片段通配。 */
  mediaTypes?: readonly string[]
  /** 文件必须同时具备这里列出的所有 capability。 */
  requiredCapabilities?: readonly FileCapability[]
  /** 文件 properties 中必须存在且严格相等的标量属性。 */
  properties?: Readonly<Record<string, EngineMatchPropertyValue>>
}>

export type EngineDescriptor = Readonly<{
  engineId: string
  label: string
  match?: EngineMatcher
  /** 数字越大，未配置用户偏好时越优先；缺省为 0。 */
  priority?: number
  layout: EngineLayout
  access: EngineAccess
  supportsStandaloneWindow?: boolean
  iconHint?: string
}>

File: native/crates/ideall-protocol/src/engine.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::FileKind;

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum EngineLayout {
    Padded,
    Fill,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum EngineAccess {
    ReadOnly,
    ReadWrite,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum EngineSuspension {
    Serializable,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EngineMatcher {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub kinds: Vec<FileKind>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub media_types: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub required_capabilities: Vec<String>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub properties: BTreeMap<String, Value>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EngineDescriptor {
    pub engine_id: String,
    pub label: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub r#match: Option<EngineMatcher>,
    #[serde(default)]
    pub priority: i32,
    pub layout: EngineLayout,
    pub access: EngineAccess,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub suspension: Option<EngineSuspension>,
    #[serde(default)]
    pub supports_standalone_window: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub icon_hint: Option<String>,
}
